lib/ddtrace/configuration/components.rb

```diff
@@ -54,28 +54,116 @@ def build_tracer(settings, agent_settings)
           tracer = settings.tracer.instance
           return tracer unless tracer.nil?
 
-          tracer = Tracer.new(
+          # Apply test mode settings if test mode is activated
+          if settings.test_mode.enabled
+            context_flush = build_test_mode_context_flush(settings)
+            sampler = build_test_mode_sampler
+            writer = build_test_mode_writer(settings, agent_settings)
+          else
+            context_flush = build_context_flush(settings)
+            sampler = build_sampler(settings)
+            writer = build_writer(settings, agent_settings)
+          end
+
+          subscribe_to_writer_events!(writer, sampler)
+
+          Tracer.new(
             default_service: settings.service,
             enabled: settings.tracer.enabled,
-            partial_flush: settings.tracer.partial_flush.enabled,
+            context_flush: context_flush,
+            sampler: sampler,
+            writer: writer,
             tags: build_tracer_tags(settings),
-            sampler: PrioritySampler.new(
+          )
+        end
+
+        def build_context_flush(settings)
+          if settings.tracer.partial_flush.enabled
+            Datadog::ContextFlush::Partial.new(min_spans_before_partial_flush: settings.tracer.partial_flush.min_spans_threshold)
+          else
+            Datadog::ContextFlush::Finished.new
+          end
+        end
+
+        # TODO: Sampler should be a top-level component.
+        # It is currently part of the Tracer initialization
+        # process, but can take a variety of options (including
+        # a fully custom instance) that makes the Tracer
+        # initialization process complex.
+        def build_sampler(settings)
+          if (sampler = settings.tracer.sampler)
+            if settings.tracer.priority_sampling == false
+              sampler
+            else
+              ensure_priority_sampling(sampler)
+            end
+          elsif settings.tracer.priority_sampling == false
+            Sampling::RuleSampler.new(
+              rate_limit: settings.sampling.rate_limit,
+              default_sample_rate: settings.sampling.default_rate
+            )
+          else
+            PrioritySampler.new(
               base_sampler: AllSampler.new,
               post_sampler: Sampling::RuleSampler.new(
                 rate_limit: settings.sampling.rate_limit,
                 default_sample_rate: settings.sampling.default_rate
               )
             )
+          end
+        end
+
+        def ensure_priority_sampling(sampler = nil)
+          if sampler.is_a?(PrioritySampler)
+            sampler
+          else
+            PrioritySampler.new(
+              base_sampler: sampler,
+              post_sampler: Sampling::RuleSampler.new
+            )
```

Comment:
I think the `RuleSampler.new` call on line 122 also should get the `rate_limit` and `default_sample_rate` arguments from `settings`, as otherwise it's a bit weird that on some codepaths we use the `settings` argument, and on others, the instance directly reaches in `Datadog.configuration` (like that class is doing).

Once we do so, we should have three initializations of `RuleSampler` that look exactly the same, so we get a free cleanup :)

Finally -- does it really make sense to have the `RuleSampler` class by default reach into `Datadog.configuration` like it is doing right now? Would it make sense for 1.0 to just make `rate_limit:` and `default_sample_rate:` required arguments with no defaults instead? (I guess I don't quite understand if that class is expected to be usable by customers directly, and if in that case it makes sense to have the nice defaults).

Reply:
> I think the RuleSampler.new call on line 122 also should get the rate_limit and default_sample_rate arguments from settings, as otherwise it's a bit weird that on some codepaths we use the settings argument, and on others, the instance directly reaches in Datadog.configuration (like that class is doing).

Good call.

> Finally -- does it really make sense to have the RuleSampler class by default reach into Datadog.configuration like it is doing right now? Would it make sense for 1.0 to just make rate_limit: and default_sample_rate: required arguments with no defaults instead? (I guess I don't quite understand if that class is expected to be usable by customers directly, and if in that case it makes sense to have the nice defaults).

The issue is that we told plenty of clients to use:
```ruby
Datadog.configure do |c|
  c.tracer sampler: Datadog::PrioritySampler.new(
    post_sampler: Datadog::Sampling::RuleSampler.new
  )
end
```
to configure [TwL](https://docs.datadoghq.com/tracing/trace_search_and_analytics/#tracing-without-limits-recommended). And this API, without explicit constructions parameters was a product requirement at the time. It's  not great in hindsight, but I fear it would break too many configuration setups today.

Reply:
> The issue is that we told plenty of clients to use (...)

Ahh, yes, reality sets in. Agree with you here, probably not worth the user pain just so that we can "make it prettier" on our side.